### ISSUE TITLE
Memory leak when running test49 introduced by patch 8.2.0823

### ISSUE THREAD

**Issue by dpelle** [vim/vim#6180]
Running test49.out with the leak sanitizer shows memory leaks in vim-8.2.0877 and older:
```
$ git diff
diff --git a/src/Makefile b/src/Makefile
index 520ef0b93..9521379e4 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -702,6 +702,7 @@ LINT_OPTIONS = -beprxzF
 # When running tests output can be found in testdir/asan.*
 #SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
 #SANITIZER_CFLAGS = -g -O0 -fsanitize=undefined -fno-omit-frame-pointer
+SANITIZER_CFLAGS = -g -O0 -fsanitize=leak -fno-omit-frame-pointer
 SANITIZER_LIBS = $(SANITIZER_CFLAGS)
 
 # MEMORY LEAK DETECTION
@@ -709,7 +710,7 @@ SANITIZER_LIBS = $(SANITIZER_CFLAGS)
 # Configuration is in the .ccmalloc or ~/.ccmalloc file.
 # Doesn't work very well, since memory linked to from global variables
 # (in libraries) is also marked as leaked memory.
-#LEAK_CFLAGS = -DEXITFREE
+LEAK_CFLAGS = -DEXITFREE
 #LEAK_LIBS = -lccmalloc
 
 # Uncomment this line to have Vim call abort() when an internal error is
```
```
$ export ASAN_OPTIONS="abort_on_error=0:detect_leaks=1"
$ ./configure --with-features=huge --enable-gui=none
$ make -j8
$ cd src/testdir
$ make test49.out
...snip...
Press ENTER or type command to continue

=================================================================
==32397==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f289506417d in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xf17d)
    #1 0x560ef5c105f7 in lalloc /home/pel/sb/vim/src/misc2.c:925
    #2 0x560ef5c104a2 in alloc /home/pel/sb/vim/src/misc2.c:828
    #3 0x560ef5c10b2f in vim_strsave /home/pel/sb/vim/src/misc2.c:1281
    #4 0x560ef5ca5679 in estack_sfile /home/pel/sb/vim/src/scriptfile.c:129
    #5 0x560ef5b9a8ed in cause_errthrow /home/pel/sb/vim/src/ex_eval.c:293
    #6 0x560ef5d9821d in emsg_core /home/pel/sb/vim/src/message.c:650
    #7 0x560ef5d9847e in emsg /home/pel/sb/vim/src/message.c:748
    #8 0x560ef5b7616d in ex_unletlock /home/pel/sb/vim/src/evalvars.c:1445
    #9 0x560ef5b75ea2 in ex_unlet /home/pel/sb/vim/src/evalvars.c:1373
    #10 0x560ef5b8f353 in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2511
    #11 0x560ef5b8c5f5 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:979
    #12 0x560ef5ca8036 in do_source /home/pel/sb/vim/src/scriptfile.c:1401
    #13 0x560ef5ca7381 in cmd_source /home/pel/sb/vim/src/scriptfile.c:950
    #14 0x560ef5ca73cd in ex_source /home/pel/sb/vim/src/scriptfile.c:976
    #15 0x560ef5b8f353 in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2511
    #16 0x560ef5b8c5f5 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:979
    #17 0x560ef5b8bb72 in do_cmdline_cmd /home/pel/sb/vim/src/ex_docmd.c:590
    #18 0x560ef5d941d8 in exe_commands /home/pel/sb/vim/src/main.c:3144
    #19 0x560ef5d90d2a in vim_main2 /home/pel/sb/vim/src/main.c:789
    #20 0x560ef5d906a0 in main /home/pel/sb/vim/src/main.c:438
    #21 0x7f28932adb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

SUMMARY: LeakSanitizer: 16 byte(s) leaked in 1 allocation(s).

shell returned 23
...snip...
Press ENTER or type command to continue

=================================================================
==322==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 27 byte(s) in 1 object(s) allocated from:
    #0 0x7f846e8f017d in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xf17d)
    #1 0x56126a4425f7 in lalloc /home/pel/sb/vim/src/misc2.c:925
    #2 0x56126a4424a2 in alloc /home/pel/sb/vim/src/misc2.c:828
    #3 0x56126a4d7716 in estack_sfile /home/pel/sb/vim/src/scriptfile.c:146
    #4 0x56126a3cc8ed in cause_errthrow /home/pel/sb/vim/src/ex_eval.c:293
    #5 0x56126a5ca21d in emsg_core /home/pel/sb/vim/src/message.c:650
    #6 0x56126a5ca576 in semsg /home/pel/sb/vim/src/message.c:776
    #7 0x56126a55a5cd in emsg_funcname /home/pel/sb/vim/src/userfunc.c:543
    #8 0x56126a55a8af in get_func_tv /home/pel/sb/vim/src/userfunc.c:615
    #9 0x56126a5625d8 in ex_call /home/pel/sb/vim/src/userfunc.c:3810
    #10 0x56126a3c1353 in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2511
    #11 0x56126a3be5f5 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:979
    #12 0x56126a55c517 in call_user_func /home/pel/sb/vim/src/userfunc.c:1376
    #13 0x56126a55cb73 in call_user_func_check /home/pel/sb/vim/src/userfunc.c:1517
    #14 0x56126a55da3b in call_func /home/pel/sb/vim/src/userfunc.c:1964
    #15 0x56126a55a855 in get_func_tv /home/pel/sb/vim/src/userfunc.c:606
    #16 0x56126a5625d8 in ex_call /home/pel/sb/vim/src/userfunc.c:3810
    #17 0x56126a3c1353 in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2511
    #18 0x56126a3be5f5 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:979
    #19 0x56126a55c517 in call_user_func /home/pel/sb/vim/src/userfunc.c:1376
    #20 0x56126a55cb73 in call_user_func_check /home/pel/sb/vim/src/userfunc.c:1517
    #21 0x56126a55da3b in call_func /home/pel/sb/vim/src/userfunc.c:1964
    #22 0x56126a55a855 in get_func_tv /home/pel/sb/vim/src/userfunc.c:606
    #23 0x56126a5625d8 in ex_call /home/pel/sb/vim/src/userfunc.c:3810
    #24 0x56126a3c1353 in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2511
    #25 0x56126a3be5f5 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:979
    #26 0x56126a4da036 in do_source /home/pel/sb/vim/src/scriptfile.c:1401
    #27 0x56126a4d9381 in cmd_source /home/pel/sb/vim/src/scriptfile.c:950
    #28 0x56126a4d93cd in ex_source /home/pel/sb/vim/src/scriptfile.c:976
    #29 0x56126a3c1353 in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2511

SUMMARY: LeakSanitizer: 27 byte(s) leaked in 1 allocation(s).

shell returned 23

Press ENTER or type command to continue
...snip...
Press ENTER or type command to continue
-----------------------------------------------------
Suppressions used:
  count      bytes template
      7        652 libXt.so.6
-----------------------------------------------------

# For flaky tests retry one time.  No tests at the moment.
#@/bin/sh -c "if test -f test.out -a test49 = test61; then \
#	  if diff test.out test49.ok; \
#	  then echo flaky test ok first time; \
#	  else rm -rf test49.failed test.out X* viminfo; \
#		VIMRUNTIME=../../runtime  ../vim -f  -u unix.vim -U NONE --noplugin --not-a-term -s dotest.in test49.in; \
#	  fi \
#	fi"
# Check if the test.out file matches test.ok.
rm -rf X* test.ok viminfo
```

The test passes despite detecting leaks.

Doing a git bisection, I see that there is no leak in
vim-8.2.0822 and the leak starts in vim-8.2.0823. So this
change introduced the leak:
```
commit 25e0f5863e9010a75a1ff0d04e8f886403968755
Author: Bram Moolenaar <Bram@vim.org>
Date:   Mon May 25 22:36:50 2020 +0200

    patch 8.2.0823: Vim9: script reload test is disabled
    
    Problem:    Vim9: script reload test is disabled.
    Solution:   Compile a function in the context of the script where it was
                defined.  Set execution stack for compiled function.  Add a test
                that an error is reported for the right file/function.
```

**Comment by brammool** [comment]
Dominique wrote:

> Running test49.out with the leak sanitizer shows memory leaks in vim-8.2.0877 and older:
> ```
> $ git diff
> diff --git a/src/Makefile b/src/Makefile
> index 520ef0b93..9521379e4 100644
> --- a/src/Makefile
> +++ b/src/Makefile
> @@ -702,6 +702,7 @@ LINT_OPTIONS = -beprxzF
>  # When running tests output can be found in testdir/asan.*
>  #SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
>  #SANITIZER_CFLAGS = -g -O0 -fsanitize=undefined -fno-omit-frame-pointer
> +SANITIZER_CFLAGS = -g -O0 -fsanitize=leak -fno-omit-frame-pointer
>  SANITIZER_LIBS = $(SANITIZER_CFLAGS)
>  
>  # MEMORY LEAK DETECTION
> @@ -709,7 +710,7 @@ SANITIZER_LIBS = $(SANITIZER_CFLAGS)
>  # Configuration is in the .ccmalloc or ~/.ccmalloc file.
>  # Doesn't work very well, since memory linked to from global variables
>  # (in libraries) is also marked as leaked memory.
> -#LEAK_CFLAGS = -DEXITFREE
> +LEAK_CFLAGS = -DEXITFREE
>  #LEAK_LIBS = -lccmalloc
>  
>  # Uncomment this line to have Vim call abort() when an internal error is
> 
> $ export ASAN_OPTIONS="abort_on_error=0:detect_leaks=1"
> $ ./configure --with-features=huge --enable-gui=none
> $ make -j8
> $ cd src/testdir
> $ make test49.out
> ...snip...
> Press ENTER or type command to continue
> 
> =================================================================
> ==32397==ERROR: LeakSanitizer: detected memory leaks
> 
> Direct leak of 16 byte(s) in 1 object(s) allocated from:
>     #0 0x7f289506417d in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xf17d)
>     #1 0x560ef5c105f7 in lalloc /home/pel/sb/vim/src/misc2.c:925
>     #2 0x560ef5c104a2 in alloc /home/pel/sb/vim/src/misc2.c:828
>     #3 0x560ef5c10b2f in vim_strsave /home/pel/sb/vim/src/misc2.c:1281
>     #4 0x560ef5ca5679 in estack_sfile /home/pel/sb/vim/src/scriptfile.c:129
>     #5 0x560ef5b9a8ed in cause_errthrow /home/pel/sb/vim/src/ex_eval.c:293
>     #6 0x560ef5d9821d in emsg_core /home/pel/sb/vim/src/message.c:650
>     #7 0x560ef5d9847e in emsg /home/pel/sb/vim/src/message.c:748
>     #8 0x560ef5b7616d in ex_unletlock /home/pel/sb/vim/src/evalvars.c:1445
>     #9 0x560ef5b75ea2 in ex_unlet /home/pel/sb/vim/src/evalvars.c:1373

[...]

Took some debugging and bisecting to find where this happens.  It's an
:unlet command that throws to errors, and the script file name in the
second wasn't freed.  Patch 8.2.0883 should fix it, please check.

-- 
hundred-and-one symptoms of being an internet addict:
250. You've given up the search for the "perfect woman" and instead,
     sit in front of the PC until you're just too tired to care.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
This issue was fixed in patch 8.2.0883. Closing.
